internal/controller/apiextensions/definition/composed.go

```diff
@@ -0,0 +1,256 @@
+	"strings"
+	"sync"
+
+	"github.com/google/uuid"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	kunstructured "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kcache "k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	cache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	runtimeevent "sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/crossplane/crossplane-runtime/pkg/controller"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+type composedResourceInformers struct {
+	log     logging.Logger
+	cluster cluster.Cluster
+
+	gvkRoutedCache *controller.GVKRoutedCache
+
+	lock      sync.RWMutex // everything below is protected by this lock
+	informers map[schema.GroupVersionKind]mrInformer
+	sinks     map[string]func(ev runtimeevent.UpdateEvent) // by some uid
+	xrs       map[string]xr                                // by controller name
+}
+
+type mrInformer struct {
+	cache    cache.Cache
+	cancelFn context.CancelFunc
+}
+
+type xr struct {
+	gvk   schema.GroupVersionKind
+	cache cache.Cache
+}
+
+var _ source.Source = &composedResourceInformers{}
+
+// Start implements source.Source, i.e. starting composedResourceInformers as
+// source with h as the sink of update events.
+func (i *composedResourceInformers) Start(ctx context.Context, h handler.EventHandler, q workqueue.RateLimitingInterface, ps ...predicate.Predicate) error {
+	id := uuid.New().String()
+
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.sinks[id] = func(ev runtimeevent.UpdateEvent) {
+		for _, p := range ps {
+			if !p.Update(ev) {
+				return
+			}
+		}
+		h.Update(ctx, ev, q)
+	}
+
+	go func() {
+		<-ctx.Done()
+
+		i.lock.Lock()
+		defer i.lock.Unlock()
+		delete(i.sinks, id)
+	}()
+
+	return nil
+}
+
+// RegisterXR maps a controller name to a GVK.
+func (i *composedResourceInformers) RegisterXR(name string, ca cache.Cache, gvk schema.GroupVersionKind) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	if i.xrs == nil {
+		i.xrs = make(map[string]xr)
+	}
+	i.xrs[name] = xr{
+		gvk:   gvk,
+		cache: ca,
+	}
+}
+
+// UnregisterXR removes a controller name to GVK mapping.
+func (i *composedResourceInformers) UnregisterXR(name string) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	delete(i.xrs, name)
+}
+
+func (i *composedResourceInformers) Watch(gvks ...schema.GroupVersionKind) {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	// start new informers
+	for _, gvk := range gvks {
+		if _, found := i.informers[gvk]; found {
+			continue
+		}
+
+		log := i.log.WithValues("gvk", gvk.String())
+
+		ctx, cancelFn := context.WithCancel(context.Background()) //nolint:govet // the outer context is closed by the caller
```

Comment:
Perhaps add a comment to remind future contributors to make sure they call `cancelFn()` in error cases? It could also be handy to explicitly point out that we `defer cancelFn` in the goroutine, which is why it's not being done here.

Reply:
done